pkg/addons/highavailability.go

```diff
@@ -15,85 +17,76 @@ import (
 	}
 
 	ncps, err := kubeutils.NumOfControlPlaneNodes(ctx, kcli)
 	if err != nil {
-		return false, errors.Wrap(err, "check control plane nodes")
+		return false, "", errors.Wrap(err, "check control plane nodes")
 	}
-	return ncps >= 3, nil
+	if ncps < 3 {
+		return false, "number of control plane nodes is less than 3", nil
+	}
+	return true, "", nil
 }
 
 // EnableHA enables high availability.
-func EnableHA(ctx context.Context, kcli client.Client, hcli helm.Client, isAirgap bool, serviceCIDR string, proxy *ecv1beta1.ProxySpec, cfgspec *ecv1beta1.ConfigSpec) error {
+func EnableHA(ctx context.Context, kcli client.Client, kclient kubernetes.Interface, hcli helm.Client, isAirgap bool, serviceCIDR string, proxy *ecv1beta1.ProxySpec, cfgspec *ecv1beta1.ConfigSpec) error {
 	loading := spinner.Start()
 	defer loading.Close()
 
+	loading.Infof("Enabling high availability")
 	logrus.Debugf("Enabling high availability")
 
 	if isAirgap {
-		loading.Infof("Enabling high availability")
-
-		domains := runtimeconfig.GetDomains(cfgspec)
-
-		// TODO (@salah): add support for end user overrides
-		sw := &seaweedfs.SeaweedFS{
-			ServiceCIDR:         serviceCIDR,
-			ProxyRegistryDomain: domains.ProxyRegistryDomain,
-		}
-		exists, err := hcli.ReleaseExists(ctx, sw.Namespace(), sw.ReleaseName())
+		hasMigrated, err := registry.IsRegistryHA(ctx, kcli)
 		if err != nil {
-			return errors.Wrap(err, "check if seaweedfs release exists")
-		}
-		if !exists {
+			return errors.Wrap(err, "check if registry data has been migrated")
+		} else if !hasMigrated {
 			logrus.Debugf("Installing seaweedfs")
-			if err := sw.Install(ctx, kcli, hcli, addOnOverrides(sw, cfgspec, nil), nil); err != nil {
-				return errors.Wrap(err, "install seaweedfs")
+			err = ensureSeaweedfs(ctx, kcli, hcli, serviceCIDR, cfgspec)
+			if err != nil {
+				return errors.Wrap(err, "ensure seaweedfs")
 			}
 			logrus.Debugf("Seaweedfs installed!")
-		} else {
-			logrus.Debugf("Seaweedfs already installed")
-		}
 
-		// TODO (@salah): add support for end user overrides
-		reg := &registry.Registry{
-			ServiceCIDR:         serviceCIDR,
-			ProxyRegistryDomain: domains.ProxyRegistryDomain,
-			IsHA:                true,
-		}
-		logrus.Debugf("Migrating registry data")
-		if err := reg.Migrate(ctx, kcli, loading); err != nil {
-			return errors.Wrap(err, "migrate registry data")
-		}
-		logrus.Debugf("Registry migration complete!")
-		logrus.Debugf("Upgrading registry")
-		if err := reg.Upgrade(ctx, kcli, hcli, addOnOverrides(reg, cfgspec, nil)); err != nil {
-			return errors.Wrap(err, "upgrade registry")
+			loading.Infof("Migrating data for high availability")
+			logrus.Debugf("Migrating data for high availability")
+			err = migrateRegistryData(ctx, kcli, kclient, cfgspec, loading)
+			if err != nil {
+				return errors.Wrap(err, "migrate registry data")
+			}
+			logrus.Debugf("Data migration complete!")
+
+			loading.Infof("Enabling registry high availability")
```

Comment:
i don't think we want to log this to the user. i think the previous log message is sufficient for this step as well

Reply:
It sits at that 100% for a long time. Are you sure there is no other message we want to log?

Reply:
i see. i would ask Alex what the copy should be then

Reply:
@ajp-io ?